feat: log deprecation warning if svelte field causes difference in resolve

dominikg opened the pull request:
- [ ] add documentation and update log message to contain link


1) do we need an option to disable these warnings
2) should they be logged upon encountering them or batched in buildEnd
3) how do we test this

dominikg commented:
example warning
```
11:46:09 AM [vite-plugin-svelte] DEPRECATION WARNING: e2e-test-dep-svelte-nested@1.0.0 package.json has `"svelte":"src/index.js"` which resolves .svelte files but standard vite resolve failed to resolve.
Using the "svelte" field in package.json is deprecated and packages should use the "svelte" exports condition instead. See :LINK HERE: for more information.
```

benmccann commented on docs/faq.md:
```diff
@@ -177,3 +177,32 @@ There is no golden rule, but you can follow these recommendations:
+
+<!-- the following header generates an anchor that is used in logging, do not modify!-->
+
+### deprecated "svelte" field
```
per @dummdidumm's point, I'm not sure we need to actually deprecate the `svelte` field. we just need to stop resolving it before `exports` and resolve it after instead. that would make the warning trigger less frequently which would be nicer for users

benmccann replied:
actually, I just looked at the code and I think it works assuming the svelte field will be left in place and only changes the order. if we really did want to get rid of it entirely we'd probably have to change the code, but I'd just suggest changing the title here and note below saying it will be removed and instead say the resolve order will change

dominikg replied:
i don't think we can recover from a resolve error unless we _always_ call `this.resolve` in a try/catch  not something i want to do long term.

dominikg replied:
also having 2 accepted ways to do this doesn't feel great. it makes writing new bundler plugins even more challanging. The only point i see in this whole excercise is that at the end of it the svelte field is no longer used at all.

benmccann replied:
There's already two ways to resolve JS files with `main` and `exports`, so it's just the analogous thing. As long as we prefer `exports` we can rip out all the code in `vite-plugin-svelte` and just defer to Vite

dominikg replied:
how do you expect
> **resolve it after instead.**

to happen after

> rip out all the code in vite-plugin-svelte and just defer to Vite

?

right now vite-plugin-svelte is a `pre` plugin exactly for the reason that we need to resolve via svelte field before vite resolve has a chance to error. Resolving the svelte field **after** a vite resolve error isn't possible unless you do in a pre-plugin. If you need to keep support for that alive after vite-plugin-svelte has switched it off, it should live in a legacy plugin that only users install that still have libraries in their project that would break otherwise. i don't want to burden regular use with that.

```js
// shortend 
async resolveId(importee,importer,options) {
  try {
    return this.resolve(importee,importer,{...options,skipSelf: true});
  } catch (e) {
    // vite resolve failed for any reason,
   const svelteResolved = await resolveViaPackageJsonSvelteField(importee,importer,options);
   if(svelteResolved) {
      log.warn('some warning here')
      return svelteResolved;
   }
   throw e;
  }
}

benmccann replied:
`vite-plugin-svelte` wouldn't resolve it at all. We've set `svelte` in [`resolve.mainFields`](https://vitejs.dev/config/shared-options.html#resolve-mainfields), so Vite will handle it.

bluwy replied:
I think we should phase out `pkg.svelte` too. `pkg.exports` is the official way to add custom condition names. Adding custom top-level fields isn't a standard and was only chosen because `pkg.exports` didn't exist then.

dominikg replied:
> `vite-plugin-svelte` wouldn't resolve it at all. We've set `svelte` in [`resolve.mainFields`](https://vitejs.dev/config/shared-options.html#resolve-mainfields), so Vite will handle it.

no it won't. if exports entry exists, mainFields are ignored. mainFields also don't resolve non-js extensions afaik. There is no way to "fall back to svelte field" without wrapping vite resolve in a try catch - at least that i know of.

benmccann replied:
> if exports entry exists, mainFields are ignored

I agree with that

> mainFields also don't resolve non-js extensions afaik

This would be news to me

benmccann replied:
> mainFields also don't resolve non-js extensions afaik

I just tested this and it appears that Vite does resolve them. In order to test I installed a component (`svelte-simple-modal`), tweaked the `package.json` in the downloaded package to only have a `svelte` field pointing to a `.svelte` file, and then commented out `vite-plugin-svelte`'s call to `resolveViaPackageJsonSvelte`. It rendered the imported component just fine

dominikg replied:
@bluwy  iirc you changed the behavior of exports + mainfields lately. Is the above still true? If yes we should be able to remove the custom code for resolveViaPackageJsonSvelte even today.

benmccann replied:
can we change the title from `deprecated "svelte" field` to something more neutral like `prefer "svelte" export condition over "svelte" field`?

> kept the svelte field as neutral instead of remove. We should not encourage them to remove it for now as that could break older tooling that still relies on it

I agree with what you said in that comment and so the word "deprecated" seems too strong to me

bluwy replied:
If `exports` is define, Vite should ignore `mainFields` (except `browser`) completely now. Vite also doesn't have a restriction on what the extensions for `mainFields` could be, so exporting a Svelte component is fine too.

benmccann replied:
@dominikg any thoughts about the heading title here? I think it might be the only outstanding item for us to resolve

dominikg replied:
> @dominikg any thoughts about the heading title here? I think it might be the only outstanding item for us to resolve

i have changed the wording to use conflicting resolve results instead of deprecation.

Rich-Harris commented on docs/faq.md:
```diff
@@ -177,3 +177,32 @@ There is no golden rule, but you can follow these recommendations:
++ "exports": {
++   "./package.json": "./package.json",
++   "./*": {
++     "svelte": "./src/*",
```
does this need to be this...

```suggestion
+     "svelte": "./src/*.svelte",
```

...in order for `library/src/Foo.svelte` to be importable as `library/Foo`? Or is that undesirable/ (Would that pattern even work?)

dominikg replied:
ideally it would be `"./*.svelte":{"svelte":"./src/*.svelte"}` but that didn't work in my testing even though node docs show a similar pattern for `*.js` 

omitting the extension altogether might confuse tools that look for it and may cause problems when there is a foo.js and a Foo.svelte  (hello windows)

bluwy replied:
I think Rich's suggestion is the correct way to support it. Using `"./*.svelte"` means they have to import as `my-lib/Foo.svelte`

dominikg replied:
i'm not an expert on building sane exports map for a new custom condition, so definetely up for improvements. Unfortunately i couldn't make the case with .svelte in the condition work, would love for someone to show a working example.

benmccann replied:
From https://github.com/nodejs/node/issues/46074#issuecomment-1370102411:

> As the creator of Rollup, I can empathise: allowing modules to import each other without specifying the .js extension was a silly mistake that added untold complexity for no good reason

If it's a bad idea for `.js` files, it's probably a bad idea for `.svelte` files too, right? I don't think I would change the example to make it work without the `.svelte` extension

dominikg replied:
So what does an ideal svelte exports condition look like today? just exporting the index.js that lists all .svelte files? Or use a glob too so that deep importing components can work?

I'd love to have a guideline for package authors here. Prebundling has made it less needed to use deep imports but they can still be a valid choice.

dominikg replied:
@dummdidumm @Rich-Harris  

ideally this faq entry contains an example of a svelte exports map that works for most/all library authors. Once we merge this PR and they start updating, many are going to have to do a breaking release. If we change the recommended setup again later that would be kind of  a d*ck move and put extra pressure on the whole ecosystem by introducing multiple major updates.

Any thoughts on how glob exports should be utilized if at all?

benmccann replied:
I've never seen anyone use glob exports in the wild before. It's probably not a good practice. E.g. if you have a main component that uses a few sub-components, you probably don't want to export the sub-components. I think I would change the example to not use glob exports

In terms of `index.js` vs deep package path, I don't know if we're wanting to make a universal declaration on that yet. You could potentially show both examples and then link to https://github.com/sveltejs/vite-plugin-svelte/blob/main/docs/faq.md#what-is-going-on-with-vite-and-pre-bundling-dependencies for a deeper discussion of what method to choose

dominikg replied:
Hadn't thought about internal subcomponents. But i do think we have to provide guidance on how to allow deep imports and/or index imports. 

svelte-package stopped generating deep exports into package.json so it would be the library authors task to remember to add new components to that. But maybe thats preferable because with glob exports everything matched by the glob is public and that makes it easier to accidentally do a breaking change without realizing.

dominikg replied:
removed the glob example and linked to the other faq entry.

bluwy commented on docs/faq.md:
```diff
@@ -177,3 +177,32 @@ There is no golden rule, but you can follow these recommendations:
+// package.json
+- "svelte": "src/index.js"
++ "exports": {
++   "./package.json": "./package.json",
```
Maybe we should not recommend export `package.json` too as it's not necessarily needed.

dominikg replied:
this is what rollup plugin svelte did recommend and before we moved it to vitefu vite-plugin-svelte still used that export to load package.json if possible. does vitefu not do it and always read from fs?

bluwy replied:
yeah vitefu will always read from the fs. I don't think there's perf issues in practice though. There's https://github.com/sveltejs/rollup-plugin-svelte/pull/198 in rollup plugin svelte to stop warning if `package.json` is not exported too.

benmccann replied:
we should definitely have the plugin read from the fs. `rollup-plugin-svelte`'s behavior is incredibly annoying. that being said, I would leave this as is for now. `rollup-plugin-svelte` still will issue a warning if it's not exported, so users should probably export it for the time being until we update `rollup-plugin-svelte`

dominikg replied:
removed the package.json line and also kept the svelte field as neutral instead of remove. We should not encourage them to remove it for now as that could break older tooling that still relies on it.

dominikg commented:
I have updated the PR to latest main and refactored it to log at buildEnd.

The new log output looks like this:
```
11:03:00 PM [vite-plugin-svelte] WARNING: The following packages use a svelte resolve configuration in package.json that is going to cause problems in the future.

e2e-test-dep-svelte-exports-simple@1.0.0

Please see https://github.com/sveltejs/vite-plugin-svelte/blob/main/docs/faq.md#deprecated-svelte-field for details.
```

benmccann commented on docs/config.md:
```diff
@@ -410,3 +410,10 @@ export default {
+- **Type** `boolean`
+- **Default:** `false`
+
+  disable svelte resolve warnings
```
I don't know if we should make this option available. If we do, we should discourage its use
```suggestion
  disable svelte resolve warnings. Note: this is highly discouraged and you should instead fix these packages which will break in the future.
```

dominikg replied:
users cannot fix these packages themselves, so we should give them an option to avoid this. 

Maybe we can change it to a list so they would still see it for newly added dependencies?

The option itself is experimental so that we can adapt on feedback

benmccann replied:
it's not very noisy since it only happens once per build. and keeping the pressure on package authors seems desirable

benmccann commented on packages/e2e-tests/resolve-exports-svelte/vite.config.js:
```diff
@@ -3,5 +3,10 @@ import { svelte } from '@sveltejs/vite-plugin-svelte';
-	plugins: [svelte({ compilerOptions: { css: 'none' } })]
+	plugins: [
+		svelte({
+			compilerOptions: { css: 'none' },
+			experimental: { disableSvelteResolveWarnings: false }
```
I think we can remove this line since it's false by default
```suggestion
			compilerOptions: { css: 'none' }
```

dominikg replied:
oops, this was from my local testing, keeping it there for easier switching wont hurt. its an e2e test app, so not affecting users in any way

benmccann replied:
it confused me a bit when I first saw it because it made me think the default might be `true`, which is why I suggested potentially removing it

benmccann commented on packages/vite-plugin-svelte/src/utils/vite-plugin-svelte-cache.ts:
```diff
@@ -124,4 +140,43 @@ export class VitePluginSvelteCache {
+}
+
+/**
+ * utility to get some info from the package.json a file belongs to (closest one with a "name" set)
```
could simplify this description:
```suggestion
 * utility to get some info from the closest package.json with a "name" set
```

benmccann commented on docs/faq.md:
```diff
@@ -208,3 +208,35 @@ module.exports = {
+
+| If you see a warning logged for this when using a svelte library, please tell the library maintainers.
+
+In the past, Svelte recommended using the custom "svelte" field in package.json to allow libraries to point at .svelte source files.
```
```suggestion
In the past, Svelte recommended using the custom `svelte` field in `package.json` to allow libraries to point at `.svelte` source files.
```

benmccann commented on docs/faq.md:
```diff
@@ -208,3 +208,35 @@ module.exports = {
+> Library authors are highly encouraged to update their packages to add the new exports condition as outlined above. Check out
+> [svelte-package](https://kit.svelte.dev/docs/packaging) which already supports it.
+>
+> For backwards compatibility, you can keep the svelte field in addition to the exports condition. But make sure that both always resolve to the same files.
```
```suggestion
> For backwards compatibility, you can keep the `svelte` field in addition to the `exports` condition. But make sure that both always resolve to the same files.
```

benmccann commented on docs/faq.md:
```diff
@@ -208,3 +208,35 @@ module.exports = {
+
+In the past, Svelte recommended using the custom "svelte" field in package.json to allow libraries to point at .svelte source files.
+This field requires a custom implementation to resolve, so you have to use a bundler plugin and this plugin needs to implement resolving.
+Since then, node has added support for [conditional exports](https://nodejs.org/api/packages.html#conditional-exports), which have more generic support in bundlers and node itself. So to increase the compatibility with the wider ecosystem and reduce the implementation needs for current and future bundler plugins, it is recommended that packages use the "svelte" exports condition.
```
```suggestion
Since then, Node has added support for [conditional exports](https://nodejs.org/api/packages.html#conditional-exports), which have more generic support in bundlers and Node itself. So to increase the compatibility with the wider ecosystem and reduce the implementation needs for current and future bundler plugins, it is recommended that packages use the `svelte` exports condition.
```

benmccann commented on docs/faq.md:
```diff
@@ -208,3 +208,34 @@ module.exports = {
+
+### conflicts in svelte resolve
+
+| If you see a warning logged for this when using a svelte library, please tell the library maintainers.
```
```suggestion
| If you see a warning logged for this when using a Svelte library, please tell the library maintainers.
```
